infra/Pos/Slotting/Impl/Util.hs

```diff
@@ -17,55 +17,79 @@ import           Pos.Core.Types              (EpochIndex, SlotId (..), Timestamp
+approxSlotUsingOutdated
+    :: (MonadSlotsData m)
+    => EpochIndex
+    -> Timestamp
+    -> m SlotId
 approxSlotUsingOutdated penult t = do
-    SlottingData {..} <- getSlottingData
-    systemStart <- getSystemStart
+    systemStart <- getSystemStartM
+    sdLast      <- getNextEpochSlottingDataM
     let epochStart = esdStartDiff sdLast `addTimeDiffToTimestamp` systemStart
     pure $
         if | t < epochStart -> SlotId (penult + 1) minBound
            | otherwise      -> outdatedEpoch systemStart t (penult + 1) sdLast
   where
     outdatedEpoch systemStart (Timestamp curTime) epoch EpochSlottingData {..} =
         let duration = convertUnit esdSlotDuration
-            start = getTimestamp (esdStartDiff `addTimeDiffToTimestamp` systemStart) in
+            start = getTimestamp (esdStartDiff `addTimeDiffToTimestamp` systemStart)
+        in
         unflattenSlotId $
         flattenEpochIndex epoch + fromIntegral ((curTime - start) `div` duration)
 
+-- TODO(ks): Move to MonadSlotsData?
 -- | Compute current slot from current timestamp based on data
 -- provided by 'MonadSlotsData'.
 slotFromTimestamp
-    :: MonadSlotsData m
-    => Timestamp -> m (Maybe SlotId)
+    :: (MonadSlotsData m, MonadThrow m)
+    => Timestamp
+    -> m (Maybe SlotId)
 slotFromTimestamp approxCurTime = do
-    SlottingData {..} <- getSlottingData
-    systemStart <- getSystemStart
-    let tryEpoch = computeSlotUsingEpoch systemStart approxCurTime
-    let penultRes = tryEpoch sdPenultEpoch sdPenult
-    let lastRes = tryEpoch (succ sdPenultEpoch) sdLast
-    return $ penultRes <|> lastRes
 
-computeSlotUsingEpoch
-    :: Timestamp
-    -> Timestamp
-    -> EpochIndex
-    -> EpochSlottingData
-    -> Maybe SlotId
-computeSlotUsingEpoch systemStart (Timestamp curTime) epoch EpochSlottingData {..}
-    | curTime < epochStart = Nothing
-    | curTime < epochStart + epochDuration = Just $ SlotId epoch localSlot
-    | otherwise = Nothing
+    systemStart     <- getSystemStartM
+    allEpochIndex   <- getAllEpochIndexM
+
+    -- We partially apply the function and then iterate over all epoch indexes.
+    mSlotId         <- forM allEpochIndex (findSlot systemStart approxCurTime)
+
+    pure $ head $ catMaybes mSlotId
+
   where
-    epochStart = getTimestamp (esdStartDiff `addTimeDiffToTimestamp` systemStart)
-    localSlotNumeric = fromIntegral $ (curTime - epochStart) `div` slotDuration
-    localSlot =
-        leftToPanic "computeSlotUsingEpoch: " $
-        mkLocalSlotIndex localSlotNumeric
 
-    slotDuration, epochDuration :: Microsecond
-    slotDuration = convertUnit esdSlotDuration
-    epochDuration = slotDuration * fromIntegral C.epochSlots
+    findSlot
+        :: (MonadSlotsData m, MonadThrow m)
+        => Timestamp
+        -> Timestamp
+        -> EpochIndex
+        -> m (Maybe SlotId)
+    findSlot systemStart approxCurTime' epochIndex = do
+        epochSlotData <- getEpochSlottingDataEmpatically epochIndex
```

Comment:
I just moved the function locally. Removed the extra function, it does simplify things a bit.